src/pymoca/tree.py

```diff
@@ -445,8 +445,138 @@ def _find_rest_of_name(
+    # declared named *component* elements of the component".
+    # This can include inherited and imported components.
+    # Look up the type (Class) within the current scope if necessary
+    found = find_name(first_name, scope, search_parent=False, copy=False)
+    if isinstance(found, ast.Symbol):
+        if next_names:
+            if isinstance(found.type, ast.ComponentRef):
+                type_name = str(found.type)
+                found_type_class = find_name(type_name, scope, copy=False)
+                if found_type_class is None or isinstance(found_type_class, ast.Symbol):
+                    scope_full_reference = str(scope.full_reference())
+                    raise NameLookupError(
+                        f'Symbol type "{type_name}" not found in scope "{scope_full_reference}"'
+                    )
+            else:
+                # type is already an InstanceClass
+                found_type_class = found.type
+            # Look in symbols of the type
+            found = _find_composite_name_in_symbols(next_names, found_type_class)
+    else:
+        found = None
+    return found
+
+
+def _find_composite_name_in_classes(name: str, scope: ast.Class) -> Optional[ast.Class]:
+    """Search for composite name (e.g. A.B.C) in local classes, recursively"""
+    first_name, _, next_names = name.partition(".")
+    found = None
+    if first_name in scope.classes:
+        found = scope.classes[first_name]
+    if found and next_names:
+        found = _find_composite_name_in_classes(next_names, found)
+    return found
+
+
+def _flatten_first_and_find_rest(
+    first: ast.Class, rest_of_name: str
+) -> Optional[Union[ast.Class, ast.Symbol]]:
+    """Lookup the `B.C` part of Composite Name Lookup (`A.B.C`) where`A` is a Class"""
+
+    # 3. If `A` is a Class:
+    #     1. `A` is temporarily flattened without modifiers
+    #     2. `B.C` is looked up among named elements of temp flattened class,
+    #     but if `A` is not a package, lookup is restricted to `encapsulated` elements only
+    #     and "the class we look inside shall not be partial in a simulation model".
+
+    #     Checking "the class we look inside shall not be partial in a simulation model"
+    #     is left to the caller.
+
+    # Spec Section 5.3.2, 4th bullet (`A` is a class):
+    # "If the identifier denotes a class, that class is temporarily
+    # flattened (as if instantiating a component without modifiers of this
+    # class, see section 7.2.2) and using the enclosing classes of the
+    # denoted class. The rest of the name (e.g., B or B.C) is looked up
+    # among the declared named elements of the temporary flattened class. If
+    # the class does not satisfy the requirements for a package, the lookup
+    # is restricted to encapsulated elements only. The class we look inside
+    # shall not be partial in a simulation model."
+    # Why do we have to temporarily flatten the class? Flattening requires name lookup
```

Comment:
Ahh, but it actually says "without modifiers" (comment on line 583)!

Here is a breadcrumb to start the research: https://github.com/modelica/ModelicaSpecification/issues/862

Reply:
I'm going to go ahead and resolve this since we already have a TODO for this. We should still be able to see the resolved comment for the breadcrumb I left above. Also, it would be nice to look for cases that fail without temporary flattening as you suggest. I expect if there are cases, we should find them in the ModelicaCompliance flattening tests and/or when we start testing against more open source libraries.